File: runtime/peerbot/src/com/newbound/p2p/protocol/SEND_READKEY.java
package com.newbound.p2p.protocol;

import org.json.JSONObject;

import com.newbound.p2p.Codes;
import com.newbound.p2p.P2PPeer;
import com.newbound.p2p.P2PService;
import com.newbound.p2p.P2PSocket;
import com.newbound.robot.Callback;

public class SEND_READKEY implements Callback {

	P2PService P2P;

	public SEND_READKEY(P2PService p2p) 
	{
		super();
		P2P = p2p;
	}

	@Override
	public void execute(final JSONObject data) 
	{
		try
		{
			String uuid = data.getString("uuid");
			System.out.println("Got read key request from "+uuid);
			P2PPeer p = P2P.getPeer(uuid);
			if (p.getPublicKey() == null) 
			{
				P2PSocket sock = (P2PSocket)data.get("p2psocket");
	
				if (!sock.CONNECTING) P2P.send(uuid, "".getBytes(), Codes.SEND_PUBKEY);
				sock.CONNECTING = true;
				
				System.out.println("Deferring read key request "+uuid);
				
				P2P.setTimeout(new Runnable() 
				{
					@Override
					public void run() 
					{
						execute(data);
					}
				}, "P2PSocket connecting, waiting to execute command", 100);
			}
			else
			{
				byte[] ba2 = p.getWriteKey().toBytes(); // Their read key is my write key
				if (!P2P.send(uuid, P2P.encryptWithPrivateKey(uuid, ba2, 0, ba2.length), Codes.READKEY)) 
					System.out.println("Unable to send read key "+uuid);
				else System.out.println("Read key sent to "+uuid);
			}
		}
		catch (Exception x)
		{
			x.printStackTrace();
		}
	}
}

File: runtime/peerbot/src/newbound/p2p/protocol/sendreadkey.py
import json
from newbound.p2p.p2presponse import P2PResponse
import newbound.p2p.codes as Codes

class SEND_READKEY(object):
    def __init__(self, service):
        self.service = service

    def execute(self, data):
        uuid = data['uuid']
        ba = data['data']
        sock = data['p2psocket']
        p = self.service.p2p.getPeer(uuid)
        print('Got read key request from '+uuid)
        sock.connecting = True
        if p.pub == None:
            #if not sock.connecting: # FIXME - this "if" may be why we need pubkey:22
            print('Cannot send read key without their public key, requesting from '+uuid)
            sock.parser.send(P2PResponse(Codes.SEND_PUBKEY, b''))
        else:
            # My write key is their read key
            ba = self.service.p2p.encryptWithPrivateKey(uuid, p.writekey.key)
            sock.parser.send(P2PResponse(Codes.READKEY, ba))
